### DISCUSSION TITLE
Executing script while navigating.

### DISCUSSION THREAD

**Discussion by christianMiguez** [vercel/next.js#51046] [Help] [answered]
### Summary

Hi guys, can you help me with the following?

I'm generating static sites from WordPress content, and some post has a script in its content:

```
<p>This is the main content of the blog post, blah blah blah blah</p>.

<script>console.log('Activated')</script>
```

I'm inyecting this html on nextjs using dangerouslySetInnerHTML:

```
<div
  className="wordpress-content"
  id="wordpress-content"
  dangerouslySetInnerHTML={{ __html: fetchApiResponse }}
/>

```
**This console log is triggered when loading the first page, but not when using the next navigation.** If I navigate to another page and then go to the same page this console.log is not triggered.

It is as if the <script> tags of the content are not read when navigating, but yes on the first load or when event "window.onload" happen.

Any of you guys has faced with that? how can i solve it? The idea is trigger the console.log() on every navigation not only page load.

Thanks in advance!

**Comment by sahariar-safin** [comment]
The problem is that the `dangerouslySetInnerHTML` prop is only evaluated when the page is first loaded. When you navigate to another page, the HTML is not re-evaluated, so the `script` tag is not executed.

To fix this, you can use the `useEffect` hook to evaluate the `dangerouslySetInnerHTML` prop on every navigation. The `useEffect` hook takes two arguments: a function and a dependency array. The function will be executed whenever any of the dependencies change. In this case, the dependency array will be the `fetchApiResponse` variable. This will ensure that the `dangerouslySetInnerHTML` prop is evaluated whenever the `fetchApiResponse` variable changes, which will happen on every navigation.

Here is an example of how to use the `useEffect` hook to solve this problem:

```jsx
import React, { useEffect } from "react";

const Page = () => {
  const [fetchApiResponse, setFetchApiResponse] = useState("");

  useEffect(() => {
    fetch("https://example.com/api/v1/posts")
      .then((response) => response.json())
      .then((data) => {
        setFetchApiResponse(data);
      });
  }, []);

  return (
    <div className="wordpress-content" id="wordpress-content">
      {fetchApiResponse}
    </div>
  );
};

export default Page;
```

This code will evaluate the `dangerouslySetInnerHTML` prop on every navigation. This will ensure that the `script` tag is executed on every navigation.

I hope this helps!

**Reply by christianMiguez** [reply]
Thanks @shafinthedeveloper for your comment, make sense what you said. The thing is that I need to convert to _fetchApiResponse_ to html because this data is string.

Also I need to make sure the content is SEO friendly.

Thanks again :)

**Comment by icyJoseph** [comment, marked as answer]
Hi,

This is normal behaviour, setting scripts with innerHTML doesn't work:

```html
<script>
  console.log("I execute")
</script>

<div id="root"></div>
```

And then load this JS

```js
const root = document.getElementById("root")
root.innerHTML = "<script>console.log('Activated')</script>";
```

It won't log `'Activated'`, only `'I execute'`.

So what's going on here...

First, some literature: 

- https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML#security_considerations

> Although this may look like a [cross-site scripting](https://en.wikipedia.org/wiki/Cross-site_scripting) attack, the result is harmless. HTML specifies that a [<script>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script) tag inserted with innerHTML [should not execute](https://www.w3.org/TR/2008/WD-html5-20080610/dom.html#innerhtml0).

Alright, so as said, this is working as expected. Why does it execute once though.

Let's say you have `/` and `/post`, `/post` has the script.

- Navigating directly to `/post` from an empty tab, for example, queries the server for HTML, and this HTML contains the script, because it was SSR'd, so the browser finds it and executes it
- Now you navigate to `/`, and back to `/post`, in this process the Router cleans up post components and renders landing page components, but when you go back to `/post`, it generated the page as part of CSR, not SSR, so the script is created with innerHTML using JavaScript through React. The browser won't execute that. 
- It follows that, arriving to `/` as entry point, and then going to `/post` also won't execute the script.

Since I imagine this content comes from a CMS of sorts. I think you have to manually parse the HTML and append it to the target div. 

For example:

```html
<div id="root"></div>
```

```js
const source = `<p>This is the main content of the blog post, blah blah blah blah</p>.

<script>console.log('Activated')</script>`

const parsed = document.createRange().createContextualFragment(source);

const root = document.getElementById("root")

root.appendChild(parsed);
```

And you'd do this inside a useEffect, etc, but the problem is that it won't be part of the initial HTML, so I think you need do the `dangerouslySetInnerHTML`, and then inside a useEffect, clear the contents of your `div` with id, `"wordpress-content"`, parse the given string with the example above, and re-append it to the `"wordpress-content"` div, but that might activate it twice on first load, haha oh boy, you are in for a ride. 

Alternatively the above might work if SEO is not a big concern, then you fetch the HTML fragment client side and append it as shown.

**Reply by christianMiguez** [reply]
Hi @icyJoseph, looks like your proposal using document.createRange() and appending parsed data works, also what you said about script being evaluated twice is happen too. 

So I'm trying something like the following:

```
const [isFirstLoad, setIsFirstLoad] = useState(true);


useEffect(() => {

  if (!isFirstLoad) {
      const parsed = document.createRange().createContextualFragment(source);
      const root = document.getElementById("root")
      root.appendChild(parsed);
    } else {
      setIsFirstLoad(false);
    }

  }, [source]);
```

So it should happen only in the first load and get it working as expected :)

Thanks

**Reply by icyJoseph** [reply]
Awesome! Please don't forget to mark as answered 😊
